### ISSUE TITLE
CUDA OOM while saving model (trying to allocate 1EB memory)

### ISSUE THREAD

**Issue by vivekmadan2** [facebookresearch/llama-recipes#45]
When training with FSDP without LORA, training finishes successfully but while saving the model, it runs out of CUDA memory. 

Error thrown by https://github.com/facebookresearch/llama-recipes/blob/e5970e2e1f59eb2af513ea269823a3b9bb8f3cc8/utils/train_utils.py#L144

Command: 
torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py --model_name PATH/TO/7B --pure_bf16 --dist_checkpoint_root_folder model_checkpoints --dist_checkpoint_folder fine-tuned --batch_size_training 8 --micro_batch_size 8 --enable_fsdp

Stack trace:

   results = train(                                                                                                                                                                                                    
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/utils/train_utils.py", line 144, in train                                                                                          
    results = train(    
model_checkpointing.save_model_and_optimizer_sharded(model, rank, train_config)
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/utils/train_utils.py", line 144, in train
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/model_checkpointing/checkpoint_handler.py", line 112, in save_model_and_optimizer_sharded
    dist_cp.save_state_dict(
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 104, in save_state_dict
    model_checkpointing.save_model_and_optimizer_sharded(model, rank, train_config)
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/model_checkpointing/checkpoint_handler.py", line 112, in save_model_and_optimizer_sharded
    dist_cp.save_state_dict(
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 104, in save_state_dict
    dist_cp.save_state_dict(
      File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 104, in save_state_dict
model_checkpointing.save_model_and_optimizer_sharded(model, rank, train_config)
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/model_checkpointing/checkpoint_handler.py", line 112, in save_model_and_optimizer_sharded
    model_checkpointing.save_model_and_optimizer_sharded(model, rank, train_config)
  File "/home/ubuntu/efs/jumpstart/llama_finetuning/updated_finetuning/llama-recipes/model_checkpointing/checkpoint_handler.py", line 112, in save_model_and_optimizer_sharded
    dist_cp.save_state_dict(
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 104, in save_state_dict
            central_plan

**Comment by chauhang** [comment]
@vivekmadan2 Please share details of the GPUs used for above. In my tests 8xA100-40g worked for pure FSDP fine-tuning for the 7B param models

**Comment by vivekmadan2** [comment]
I am working on g5.48xlarge (8GPUs, 24GB per GPU). But, it should not take that much cuda memory to just save.

**Comment by vivekmadan2** [comment]
I am trying with lower batch size but I don't understand why it takes so much GPU memory to save the model. Why is it using extra memory to save the model?

**Comment by vivekmadan2** [comment]
@HamidShojanazeri With smaller batch size I am able to save the model for the first few times but after that it again results in the same CUDA OOM error. 

Suppose validation loss decreases after each epoch. Then, it saves the 7B model for first three epochs but for the fourth epoch it runs into CUDA OOM error. Since you have more CUDA memory, do you mind trying for 20 epochs (without LORA) in a setting where validation loss decreases in each epoch.

**Comment by vivekmadan2** [comment]
For 13B, after first epoch

CPU Total Peak Memory consumed during the train (max): 5 GB                                                                                                                                                             
Max CUDA memory allocated was 12 GB                                                                                                                                                                                     
Max CUDA memory reserved was 19 GB   

After 2nd epoch: 
CPU Total Peak Memory consumed during the train (max): 6 GB                                                                                                                                                             
Max CUDA memory allocated was 17 GB                                                                                                                                                                                     
Max CUDA memory reserved was 21 GB   

Why is there more memory being used? Some cuda memory is not freed and it keeps piling up.

**Comment by aditya-essential** [comment]
I am having a similar issue, my model training + eval runs fine for the epoch 0, then the model saves but epoch 1 fails due to OOM. 

I am using FSDP w/o PEFT on p4de.24xlarge cluster. I verified that by removing the saving step the training proceeds just fine. 

Also updated https://github.com/facebookresearch/llama-recipes/blob/main/model_checkpointing/checkpoint_handler.py#L67 to have `ShardedStateDictConfig(offload_to_cpu=True)` but the problem persists.

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 , @aditya-essential can you pls add the following to your main script. llama-finetuning.py, it might be from memory fragmentation, that this should help.

```
import os
os.env['PYTORCH_CUDA_ALLOC_CONF']='expandable_segments:True'
```

**Comment by HamidShojanazeri** [comment]
> Side question: Do you have inference script for how to load the model if it is fine-tuned without PEFT? (checkpoint format seems different than .pth format of the original model or the .bin format compatible with transformers. Created a new issue for visibility for others #53

As I mentioned, this [PR](https://github.com/facebookresearch/llama-recipes/pull/39) help to convert FSDP checkpoints into HF checkpoints then the [inference.py](https://github.com/facebookresearch/llama-recipes/blob/main/inference/inference.py) works as expected. Please look at the README in the PR.

**Comment by vivekmadan2** [comment]
@HamidShojanazeri I tried this and the error still persists for me. 

@aditya-essential Did you get rid of this error?

> @vivekmadan2 , @aditya-essential can you pls add the following to your main script. llama-finetuning.py, it most probably comes from memory fragmentation, in this case the following should help.
> 
> ```
> import os
> os.environ['PYTORCH_CUDA_ALLOC_CONF']='expandable_segments:True'
> ```

**Comment by aditya-essential** [comment]
@vivekmadan2 the advised soln doesnt work for me either.

**Comment by lwmlyy** [comment]
@aditya-essential @vivekmadan2 Did you guys solve the ckpt-saving problem?

**Comment by vivekmadan2** [comment]
No. It did not work when fine-tuning the entire model (not using LORA).

**Comment by lwmlyy** [comment]
> No. It did not work when fine-tuning the entire model (not using LORA).

It didnt work with lora either. It went CUDA OOM after training and evaluation at epoch 1.
I use the following command:
torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py --enable_fsdp --low_cpu_fsdp --model_name ../Llama-2-70b-chat-hf --micro_batch_size 1 --batch_size_training 12 --dist_checkpoint_root_folder ../Llama-2-70b-chat-hf/ --dist_checkpoint_folder fine-tuned  --use_peft --peft_method lora --lr 3e-4 --epoch 2 --pure_bf16 --dataset alpaca_dataset --output_dir llama-70b-lorawallsft

**Comment by HamidShojanazeri** [comment]
> torchrun --nnodes 1 --nproc_per_node 8 llama_finetuning.py --enable_fsdp --low_cpu_fsdp --model_name ../Llama-2-70b-chat-hf --micro_batch_size 1 --batch_size_training 12 --dist_checkpoint_root_folder ../Llama-2-70b-chat-hf/ --dist_checkpoint_folder fine-tuned --use_peft --peft_method lora --lr 3e-4 --epoch 2 --pure_bf16 --dataset alpaca_dataset --output_dir llama-70b-lorawallsft

This is expected, you would need at least 2 nodes of A100s in our experience for 70B.

**Comment by HamidShojanazeri** [comment]
closing it as the issue is stale and not able to repro anymore.
